DOCS-4422-4.1-migrate-component-authorization-and-ldap-spring-kt (#1148)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/setting-up-ldap-provider-for-spring-security.adoc

= Configure LDAP Provider for Spring Security
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: anypoint studio, esb, connector, spring security, spring, ldap, authentication

> In Mule 4, you can use a Spring Security LDAP provider to perform component authorization, or use it as a Mule security provider via the Spring Module. Before you configure the LDAP provider obtain the Spring Jar files and declare Spring Beans.

Reviewer: ```suggestion
In Mule 4, you can either use a Spring Security LDAP provider to perform component authorization, or use it as a Mule security provider via the Spring module. Before you configure the LDAP provider, obtain the Spring JAR files and declare Spring beans.
```
I currently have Spring module and Spring beans in the term presentation list. Please let me know if you have other input for Spring terminology presentation rules. I looked at this, but became overwhelmed: https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/.

== Obtain JAR Files

> The Mule software distribution provides the Spring `.jar` files you need in the `<distribution>/lib/opt` directory:

Reviewer: ```suggestion
The Mule software distribution provides the Spring JAR files you need in the `<distribution>/lib/opt` directory:
```

	* `spring-ldap-core-2.3.2.RELEASE.jar`

	* `spring-security-config-4.2.6.RELEASE.jar`

	* `spring-security-core-4.2.6.RELEASE.jar`

	* `spring-security-ldap-4.2.6.RELEASE.jar`

	* `spring-security-web-4.2.6.RELEASE.jar`

== Declare Spring Beans

> The https://docs.spring.io/spring-security/site/docs/4.0.x/apidocs/org/springframework/security/ldap/DefaultSpringSecurityContextSource.html[DefaultSpringSecurityContextSource] is the access point for obtaining an LDAP context. +

Reviewer: What is DefaultSpringSecurityContextSource? a parameter? something else?
"The DefaultSpringSecurityContextSource..." should be followed grammatically by a common noun object, such as parameter (or whatever it is). but I see that Spring uses this concatenation for its names? 
https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/

You must declare the Spring bean for `DefaultSpringSecurityContextSource` in a separate `beans.xml` file in your resources folder.

> For example: Set up an LDAP context source for use by the Spring Security authentication provider to search and authenticate your users. Also, you need to define an `authentication-manager` with an embedded `ldap-authentication-provider` as in the following code sample:

Reviewer: ```suggestion
For example, you must set up an LDAP context source for use by the Spring Security authentication provider to search and authenticate your users. Also, you need to define an `authentication-manager` with an embedded `ldap-authentication-provider`, as in the following code sample:
```

[source,xml,linenums]
----
	<beans xmlns="http://www.springframework.org/schema/beans"
	  xmlns:context="http://www.springframework.org/schema/context"
	  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	  xmlns:jdbc="http://www.springframework.org/schema/jdbc"
	  xmlns:ss="http://www.springframework.org/schema/security"
	  xsi:schemaLocation="http://www.springframework.org/schema/beans
	  ...
	  http://www.springframework.org/schema/security/spring-security-4.2.xsd">

		<bean id="ldapContextSource" class="org.springframework.security.ldap.DefaultSpringSecurityContextSource">
	        <constructor-arg value="ldap://localhost:389"/>
	        <property name="userDn" value="cn=admin,dc=example,dc=org"/>
	        <property name="password" value="admin"/>
	    </bean>

	    <ss:authentication-manager alias="ldapAuthManager">
			<ss:ldap-authentication-provider
				server-ref="ldapContextSource"
				user-search-base="DC=example,DC=org"
				user-search-filter="(uid={0})"
				group-search-base="DC=example,DC=org"
				group-search-filter="({0})"
				group-role-attribute="ou"
			/>
	    </ss:authentication-manager>

	</beans>
----

== Configure the Mule Security Provider

> The `SpringSecurityProviderAdapter` delegates to an `AuthenticationProvider` such as the `LdapAuthenticationProvider`. To configure the Mule security provider review the following example configuration that shows how you can achieve connector-level security and other security features in Mule that require one or more security providers.

Reviewer: ```suggestion
The `SpringSecurityProviderAdapter` delegates to an `AuthenticationProvider` such as the `LdapAuthenticationProvider`. To configure the Mule security provider, review the following example configuration that shows how you can achieve in Mule connector-level security and other security features that require one or more security providers.
```

	<spring:security-manager>
 		<spring:delegate-security-provider name="ldap-provider" delegate-ref="ldapAuthManager" />
	</spring:security-manager>
----

The following example configuration references a basic HTTP security filter:

[source,xml,linenums]
----
	<http:listener-config name="HTTP_Listener_config" doc:name="HTTP Listener config">
		<http:listener-connection host="0.0.0.0" port="8081" />
	</http:listener-config>

	<spring:config name="Spring_Config"  files="beans.xml" />

	<spring:security-manager>
		<spring:delegate-security-provider name="ldap-provider" delegate-ref="ldapAuthManager" />
	</spring:security-manager>

	<flow name="secureFlow">
		<http:listener doc:name="Listener" config-ref="HTTP_Listener_config" path="/test" />

		<http:basic-security-filter doc:name="Basic security filter" securityProviders="#[['ldap-provider']]" realm="mule-realm"/>

		<ee:transform doc:name="Transform Message">
			<ee:message>
				<ee:set-payload><![CDATA[%dw 2.0
					output application/json
					---
					{
						"status": "ok"
					}]]></ee:set-payload>
			</ee:message>
		</ee:transform>
	</flow>
----

== Configure the Component Authorization

> To configure the component authorization, declare the `authorization-filter` field as in the following example:

Reviewer: ```suggestion
To configure the component authorization, declare the `authorization-filter` field as follows:
```
